lib/Doctrine/ORM/Internal/CommitOrderCalculator.php

```diff
@@ -20,137 +21,163 @@
+        $vertex->dependencyList = array();
 
-        // Go
-        foreach ($this->_classes as $node) {
-            if ($this->_nodeStates[$node->name] == self::NOT_VISITED) {
-                $this->_visitNode($node);
-            }
-        }
-
-        $sorted = array_reverse($this->_sorted);
-
-        $this->_sorted = $this->_nodeStates = array();
-
-        return $sorted;
+        $this->nodeList[$hash] = $vertex;
     }
 
     /**
-     * @param \Doctrine\ORM\Mapping\ClassMetadata $node
+     * Adds a new dependency (edge) to the graph using their hashes.
+     *
+     * @param string  $fromHash
+     * @param string  $toHash
+     * @param integer $weight
      *
      * @return void
      */
-    private function _visitNode($node)
+    public function addDependency($fromHash, $toHash, $weight)
     {
-        $this->_nodeStates[$node->name] = self::IN_PROGRESS;
+        $vertex = $this->nodeList[$fromHash];
+        $edge   = new \stdClass();
 
-        if (isset($this->_relatedClasses[$node->name])) {
-            foreach ($this->_relatedClasses[$node->name] as $relatedNode) {
-                if ($this->_nodeStates[$relatedNode->name] == self::NOT_VISITED) {
-                    $this->_visitNode($relatedNode);
-                }
-            }
-        }
+        $edge->from   = $fromHash;
+        $edge->to     = $toHash;
+        $edge->weight = $weight;
 
-        $this->_nodeStates[$node->name] = self::VISITED;
-        $this->_sorted[] = $node;
+        $vertex->dependencyList[$toHash] = $edge;
     }
 
     /**
-     * @param \Doctrine\ORM\Mapping\ClassMetadata $fromClass
-     * @param \Doctrine\ORM\Mapping\ClassMetadata $toClass
+     * Return a valid order list of all current nodes.
+     * The desired topological sorting is the reverse post order of these searches.
      *
-     * @return void
-     */
-    public function addDependency($fromClass, $toClass)
-    {
-        $this->_relatedClasses[$fromClass->name][] = $toClass;
-    }
-
-    /**
-     * @param string $className
+     * {@internal Highly performance-sensitive method.}
      *
-     * @return bool
+     * @return array
      */
-    public function hasClass($className)
+    public function sort()
     {
-        return isset($this->_classes[$className]);
+        foreach ($this->nodeList as $vertex) {
+            if ($vertex->state !== self::NOT_VISITED) {
+                continue;
+            }
+
+            $this->visit($vertex);
+        }
+
+        $sortedList = $this->sortedNodeList;
+
+        $this->nodeList       = array();
+        $this->sortedNodeList = array();
+
+        return array_reverse($sortedList);
     }
 
     /**
-     * @param \Doctrine\ORM\Mapping\ClassMetadata $class
+     * Visit a given node definition for reordering.
      *
-     * @return void
+     * {@internal Highly performance-sensitive method.}
+     *
+     * @param \stdClass $vertex
      */
-    public function addClass($class)
```

Comment:
Here is a more detailed reason why I think this should not be reverted and why `data-fixtures` usage of the old class here is broken: https://github.com/doctrine/data-fixtures/pull/212#issuecomment-161023062